src/main/java/com/bookstore/repository/impl/BookRepositoryImpl.java

```diff
@@ -29,20 +33,28 @@ public Book save(Book book) {
             if (transaction != null) {
                 transaction.rollback();
             }
-            throw new RuntimeException("Cannot save book: " + book, e);
+            throw new DataProcessingException("Cannot save book: " + book, e);
         } finally {
             if (session != null) {
                 session.close();
             }
         }
     }
     
+    @Override
+    public Optional<Book> findById(Long id) {
+        try (EntityManager entityManager = sessionFactory.createEntityManager()) {

```

Comment:
you can use Session here as in other places 
```suggestion
        try (Session session = sessionFactory.openSession()) {
```